src/misc/ethernet_slirp.cpp

```diff
@@ -0,0 +1,411 @@
+{
+	(void)opaque;
+	struct timespec ts;
+	clock_gettime(CLOCK_REALTIME, &ts);
+	/* if clock_gettime fails we have more serious problems */
+	return ts.tv_nsec + (ts.tv_sec * 1e9);
+}
+
+void *slirp_timer_new(SlirpTimerCb cb, void *cb_opaque, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	return conn->TimerNew(cb, cb_opaque);
+}
+
+void slirp_timer_free(void *timer, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	struct slirp_timer *real_timer = (struct slirp_timer *)timer;
+	conn->TimerFree(real_timer);
+}
+
+void slirp_timer_mod(void *timer, int64_t expire_time, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	struct slirp_timer *real_timer = (struct slirp_timer *)timer;
+	conn->TimerMod(real_timer, expire_time);
+}
+
+int slirp_add_poll(int fd, int events, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	return conn->PollAdd(fd, events);
+}
+
+int slirp_get_revents(int idx, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	return conn->PollGetSlirpRevents(idx);
+}
+
+void slirp_register_poll_fd(int fd, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	conn->PollRegister(fd);
+}
+
+void slirp_unregister_poll_fd(int fd, void *opaque)
+{
+	SlirpEthernetConnection *conn = (SlirpEthernetConnection *)opaque;
+	conn->PollUnregister(fd);
+}
+
+void slirp_notify(void *opaque)
+{
+	(void)opaque;
+	return;
+}
+
+/* End boilerplate */
+
+SlirpEthernetConnection::SlirpEthernetConnection()
+        : EthernetConnection(),
+          config(),
+          timers(),
+          get_packet_callback(),
+          registered_fds(),
+#ifdef WIN32
+          readfds(),
+          writefds(),
+          exceptfds()
+#else /* !WIN32 */
+          polls()
+#endif
+{
+	slirp_callbacks.send_packet = slirp_receive_packet;
+	slirp_callbacks.guest_error = slirp_guest_error;
+	slirp_callbacks.clock_get_ns = slirp_clock_get_ns;
+	slirp_callbacks.timer_new = slirp_timer_new;
+	slirp_callbacks.timer_free = slirp_timer_free;
+	slirp_callbacks.timer_mod = slirp_timer_mod;
+	slirp_callbacks.register_poll_fd = slirp_register_poll_fd;
+	slirp_callbacks.unregister_poll_fd = slirp_unregister_poll_fd;
+	slirp_callbacks.notify = slirp_notify;
+}
+
+SlirpEthernetConnection::~SlirpEthernetConnection()
+{
+	if (slirp)
+		slirp_cleanup(slirp);
+}
+
+bool SlirpEthernetConnection::Initialize(Section *dosbox_config)
+{
+	Section_prop *section = static_cast<Section_prop *>(dosbox_config);
+
+	LOG_MSG("SLIRP: Slirp version: %s", slirp_version_string());
+
+	/* Config */
+	config.version = 1;
+	config.restricted = section->Get_bool("restricted");
+	config.disable_host_loopback = section->Get_bool("disable_host_loopback");
+	config.if_mtu = section->Get_int("mtu"); /* 0 = IF_MTU_DEFAULT */
+	config.if_mru = section->Get_int("mru"); /* 0 = IF_MRU_DEFAULT */
```

Comment:
MTU/MRU are maximum transmission/receive sizes. These control the maximum size of packets slirp will give to use. I don't know why users would want to change this.

Reply:
For settings, I see only benefits if we drop all these settings from the conf file and use slirp's defaults.
Because slirp's virtualizes on the software side and not at the host level, we can use all of slirp's defaults  - such as internal IPs (for the NE2000 NIC), for the internal DHCP server, and so on.
One might think that these IPs could conflict somehow - but they're not visible at the host-level or across processes. 
You can have two DOSBox instances running on the same host, using the same host's network (or 127.0.0.1) both running with the same slirp internal values, and they won't conflict.

The benefits I see with locking these down to slirp's defaults are:
 1) people writing documentation for new users can provide exact values or even drag-n'-drop config files that "just work" in staging.
 3) People packaging pre-installed retro packs can configure their software (or Windows) assuming these default IP settings, and be confident that DOSBox Staging will supprt them "out of the box", and users _won't_ be able to break their configuratios.
 4) Other DOSBox's forks will benefit in the same way, provided those users don't touch their slirp settings :-)

Reply:
Slirp doesn't provide defaults, we do. Could you explain how those benefits come from locking down defaults but would be disturbed by giving users the option to modify them?

Reply:
> Slirp doesn't provide defaults, we do.

Thanks; indeed grepping the current supported source confirms this. I was going on memory when libslirp did previously  provide defaults:

default route: 10.0.2.2/24
DNS forward: 10.0.2.3
DHCP addresses: 10.0.2.15 - 10.0.2.31

(Such as here: https://github.com/rd235/libslirp#libslirp-tutorial)

> how would those benefits be disturbed by giving users the option to modify them?

The benefit is consistency.  

DOSBox Staging will be able to provide a strong guarantee of consistent settings to those writing documentation, sharing configuration records or files, and to those pre-configuring retro software packs.

All of these activities can deliver an immutable 
result that /will just work/ on Staging.  

Those guarantees go out the window if the user (or other software packages) can modify these settings.

Reply:
I don't buy in to that. If this is required for the PR, I'm out of this.

Reply:
It's not required, and won't impact this being merged.

Reply:
My question was only about `mtu` and `mru`; if we, as maintainers, don't know about the situation when changing this would benefit users, then there's no point in making it a user-selectable option (… and writing documentation for it, passing it to translators, answering users' questions about it, etc, etc).

We don't even know what unit is it supposed to have - B, kB, KiB? :) I hope slirp does sensible thing by default here - if not we could set maximum for IPv4 networking (for MTU it's 64KiB)

Therefore I suggest to remove these two user settings. @Jookia - what do you think?

Reply:
I'm up for removing those. The value (in bytes) should be set to what the adapter can handle I guess. It might be better to pass them from the ne2000 code to the Ethernet connection on initialization. 1500 bytes is the maximum, I'm not sure where you got the 64KiB figure from.

Reply:
I agree with both of you regarding removing MRU and MTU, and setting them to 1500 bytes (which is per the Ethernet standard prior to the addition of Jumbo frames). https://tools.ietf.org/html/rfc894.

Certainly for emulating basic Ethernet in Windows 3.1 or Windows 95, there will be no need to adjust these values from 1500 bytes.

Reply:
Slirp creates a virtual subnet at the software-level for applications that use it. At the host-level, slirp is merely an application that uses the local network, similar to Firefox.

Slirp does not create a new LAN subnet at the host level.  To prove this, create two conf files:

**`15.conf`**, which gives the local DOSBox instance the IP 10.0.2.15:

``` ini
[ethernet, slirp]
restricted            = false
disable_host_loopback = false
mtu                   = 0
mru                   = 0
ipv4_network          = 10.0.2.0
ipv4_netmask          = 255.255.255.0
ipv4_host             = 10.0.2.3
ipv4_nameserver       = 10.0.2.4
ipv4_dhcp_start       = 10.0.2.15
```

**`20.conf`**, which gives the local DOSBox instance the IP 10.0.2.20:

``` ini
[ethernet, slirp]
restricted            = false
disable_host_loopback = false
mtu                   = 0
mru                   = 0
ipv4_network          = 10.0.2.0
ipv4_netmask          = 255.255.255.0
ipv4_host             = 10.0.2.5
ipv4_nameserver       = 10.0.2.6
ipv4_dhcp_start       = 10.0.2.20
```

Using the package [ethernet.zip](https://github.com/dosbox-staging/dosbox-staging/files/7644216/ethernet.zip), launch two instances of DOSBox with each:
 - `dosbox -conf 15.conf .`
 - `dosbox -conf 20.conf .`

Launch `run.bat` within both DOSBox instance to setup the slirp 10.0.2.0 "VLAN", gateway, DHCP servers, and client IPs.

Now try to ping any of the slirp IPs from the other DOSBox instance or the host. They will all fail, beacuse those Slirp IPs only exist within the software space, and aren't created at the host-level.

![2021-12-02_10-04](https://user-images.githubusercontent.com/1557255/144480442-41aacfa6-44ec-411d-977b-ad4d2de1c64c.png)
![2021-12-02_10-05](https://user-images.githubusercontent.com/1557255/144480444-3293e8a7-1672-4ac8-b2d7-fdfb631a97ec.png)
![2021-12-02_10-08](https://user-images.githubusercontent.com/1557255/144480445-2475c04a-af0e-471c-92bf-bffcab780cfa.png)
![2021-12-02_10-08_1](https://user-images.githubusercontent.com/1557255/144480448-1817d558-25fe-4fbc-8c86-fdce49466954.png)

Reply:
All comments addressed.